### PR TITLE
Rename DeviceMethod to DirectMethod

### PR DESCRIPTION
**Pull request by timtay-microsoft**
CloudToDeviceMethod -> DirectMethodRequest
CloudToDeviceMethodResult -> DirectMethodResponse

### CODE REVIEW

File: iothub/service/src/DirectMethod/DirectMethodRequest.cs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        /// </summary>
+        [JsonProperty("methodName", Required = Required.Always)]
+        public string MethodName { get; set; }
+
+        /// <summary>
+        /// The timeout (in seconds) before the direct method request will fail if the device doesn't respond to the request.
+        /// This timeout may happen if the target device is slow in handling the direct method.
+        /// </summary>
+        [JsonProperty("responseTimeoutInSeconds", NullValueHandling = NullValueHandling.Ignore)]
+        public int? ResponseTimeout { get; set; }
```
Comment (timtay-microsoft) [inline iothub/service/src/DirectMethod/DirectMethodRequest.cs:25]: These two timeout fields used to be timespans, but I don't feel that represents the field well here. The json object sent over the wire represents this timeout as an int. On top of that, making these timespans suggests a level of precision (1.15 seconds vs 1 second) that isn't possible

File: e2e/test/iothub/method/MethodE2ETests.cs [hunk 2]
```diff
@@ -256,11 +256,11 @@ await deviceClient
                     .ConfigureAwait(false);
 
                 using var serviceClient = new IotHubServiceClient(TestConfiguration.IoTHub.ConnectionString);
-                CloudToDeviceMethod c2dMethod = new CloudToDeviceMethod(commandName, TimeSpan.FromMinutes(1), TimeSpan.FromMinutes(1)).SetPayloadJson(null);
+                DirectMethodRequest c2dMethod = new DirectMethodRequest(commandName, TimeSpan.FromMinutes(1), TimeSpan.FromMinutes(1)).SetPayloadJson(null);
 
                 // act
 
-                CloudToDeviceMethodResult result = await serviceClient.DirectMethods.InvokeAsync(testDevice.Id, c2dMethod).ConfigureAwait(false);
+                DirectMethodResponse result = await serviceClient.DirectMethods.InvokeAsync(testDevice.Id, c2dMethod).ConfigureAwait(false);
```
Comment (brycewang-microsoft) [inline e2e/test/iothub/method/MethodE2ETests.cs:263]: Do we wanna use the same variable name here for all "DirectMethodResponse" object? Maybe "response" like what we did on L294

File: e2e/test/iothub/method/MethodE2ETests.cs [hunk 3]
```diff
@@ -256,11 +256,11 @@ await deviceClient
                     .ConfigureAwait(false);
 
                 using var serviceClient = new IotHubServiceClient(TestConfiguration.IoTHub.ConnectionString);
-                CloudToDeviceMethod c2dMethod = new CloudToDeviceMethod(commandName, TimeSpan.FromMinutes(1), TimeSpan.FromMinutes(1)).SetPayloadJson(null);
+                DirectMethodRequest c2dMethod = new DirectMethodRequest(commandName, TimeSpan.FromMinutes(1), TimeSpan.FromMinutes(1)).SetPayloadJson(null);
```
Comment (drwill-ms) [inline e2e/test/iothub/method/MethodE2ETests.cs:259]:
```suggestion
                var c2dMethod = new DirectMethodRequest(commandName, TimeSpan.FromMinutes(1), TimeSpan.FromMinutes(1)).SetPayloadJson(null);
```

File: e2e/test/iothub/method/MethodE2ETests.cs (hunk 2 above)
Comment (timtay-microsoft) [reply e2e/test/iothub/method/MethodE2ETests.cs:263]: Sure

File: iothub/service/src/DirectMethod/DirectMethodRequest.cs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+namespace Microsoft.Azure.Devices
+{
+    /// <summary>
+    /// Parameters to execute a direct method on the device
+    /// </summary>
+    public class DirectMethodRequest
+    {
+        /// <summary>
+        /// Method to run
```
Comment (drwill-ms) [inline iothub/service/src/DirectMethod/DirectMethodRequest.cs:15]:
```suggestion
        /// The method name to run.
```

File: iothub/service/src/DirectMethod/DirectMethodRequest.cs (hunk 1 above)
Comment (drwill-ms) [inline iothub/service/src/DirectMethod/DirectMethodRequest.cs:25]: Let's make this a TimeSpan.

https://github.com/Azure/azure-iot-sdk-csharp/blob/previews/v2/iothub/service/src/Jobs/JobRequest.cs#L64-L75

File: iothub/service/src/DirectMethod/DirectMethodRequest.cs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        /// when this direct method request is made.
+        /// </summary>
+        [JsonProperty("connectTimeoutInSeconds", NullValueHandling = NullValueHandling.Ignore)]
+        public int? ConnectionTimeout { get; set; }
+
+        [JsonProperty("payload")]
+        internal JRaw Payload { get; set; }
+
+        /// <summary>
+        /// Set payload as json
```
Comment (drwill-ms) [inline iothub/service/src/DirectMethod/DirectMethodRequest.cs:40]:
```suggestion
        /// Set the serialized JSON payload.
```

File: iothub/service/src/DirectMethod/DirectMethodRequest.cs [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+                {
+                    throw new ArgumentException(ex.Message, nameof(json)); // @ailn: here we want to hide the fact we're using Json.net
+                }
+            }
+
+            return this; // @ailn: it will allow such code: new CloudToDeviceMethod("c2dmethodname").SetPayloadJson("{ ... }");
+        }
+
+        /// <summary>
+        /// Get payload as json
```
Comment (drwill-ms) [inline iothub/service/src/DirectMethod/DirectMethodRequest.cs:65]:
```suggestion
        /// Get the serialized JSON payload.
```

File: iothub/service/src/DirectMethod/DirectMethodResponse.cs [hunk 7]
```diff
@@ -9,7 +9,7 @@ namespace Microsoft.Azure.Devices
     /// <summary>
     /// Represents the Device Method Invocation Results.
```
Comment (drwill-ms) [inline iothub/service/src/DirectMethod/DirectMethodResponse.cs:10]:
```suggestion
    /// The result of a device's direct method invocation.
```

File: iothub/service/src/DirectMethod/DirectMethodRequest.cs [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+        /// </summary>
+        public string GetPayloadAsJson()
```
Comment (drwill-ms) [inline iothub/service/src/DirectMethod/DirectMethodRequest.cs:67]: can this just be a readonly property?

Comment (drwill-ms) [reply iothub/service/src/DirectMethod/DirectMethodRequest.cs:67]: Would need a `[JsonIgnore]` attribute.

Comment (timtay-microsoft) [reply iothub/service/src/DirectMethod/DirectMethodRequest.cs:67]: Yep

File: iothub/service/src/DirectMethod/DirectMethodsClient.cs [hunk 9]
```diff
@@ -1,4 +1,6 @@
-﻿using System;
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+using System;
```
Comment (drwill-ms) [inline iothub/service/src/DirectMethod/DirectMethodsClient.cs:3]: insert a new between commend header and usings list.

File: iothub/service/src/DirectMethod/DirectMethodsClient.cs [hunk 10]
```diff
@@ -41,10 +43,10 @@ internal DirectMethodsClient(string hostName, IotHubConnectionProperties credent
         /// Invokes a method on a device.
         /// </summary>
         /// <param name="deviceId">The device identifier for the target device.</param>
-        /// <param name="cloudToDeviceMethod">Parameters to execute a direct method on the device.</param>
+        /// <param name="directMethodRequest">Parameters to execute a direct method on the device.</param>
         /// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
-        /// <returns>The <see cref="CloudToDeviceMethodResult"/>.</returns>
-        /// <exception cref="ArgumentNullException">Thrown when the provided <paramref name="deviceId"/> or <paramref name="cloudToDeviceMethod"/> is null.</exception>
+        /// <returns>The <see cref="DirectMethodResponse"/>.</returns>
```
Comment (drwill-ms) [inline iothub/service/src/DirectMethod/DirectMethodsClient.cs:48]: Don't link to the type. It is already present in the API surface and will be automatically linked.

File: iothub/service/src/DirectMethod/DirectMethodsClient.cs [hunk 11]
```diff
@@ -56,21 +58,21 @@ internal DirectMethodsClient(string hostName, IotHubConnectionProperties credent
...
+                Argument.RequireNotNull(directMethodRequest, nameof(directMethodRequest));
                 cancellationToken.ThrowIfCancellationRequested();
 
-                using HttpRequestMessage request = _httpRequestMessageFactory.CreateRequest(HttpMethod.Post, GetDeviceMethodUri(deviceId), _credentialProvider, cloudToDeviceMethod);
-                HttpResponseMessage response = await _httpClient.SendAsync(request, cancellationToken).ConfigureAwait(false);
-                await HttpMessageHelper2.ValidateHttpResponseStatusAsync(HttpStatusCode.OK, response).ConfigureAwait(false);
-                return await HttpMessageHelper2.DeserializeResponseAsync<CloudToDeviceMethodResult>(response, cancellationToken).ConfigureAwait(false);
+                using HttpRequestMessage request = _httpRequestMessageFactory.CreateRequest(HttpMethod.Post, GetDeviceMethodUri(deviceId), _credentialProvider, directMethodRequest);
+                HttpResponseMessage response = await _httpClient.SendAsync(request, cancellationToken);
+                await HttpMessageHelper2.ValidateHttpResponseStatusAsync(HttpStatusCode.OK, response);
```
Comment (drwill-ms) [inline iothub/service/src/DirectMethod/DirectMethodsClient.cs:74]: Need to add `ConfigureAwait(false)` to all awaited calls.

File: iothub/service/src/DirectMethod/DirectMethodRequest.cs (hunk 1 above)
Comment (drwill-ms) [reply iothub/service/src/DirectMethod/DirectMethodRequest.cs:25]: Also ConnectionTimeout

Comment (drwill-ms) [reply iothub/service/src/DirectMethod/DirectMethodRequest.cs:25]: I'd add the precision info in doc comments.

File: iothub/service/src/DirectMethod/DirectMethodRequest.cs [hunk 12]
```diff
@@ -18,58 +18,65 @@ public class DirectMethodRequest
...
         /// the request to the target device in time. If this value is set to 0 seconds, then the target device must be online
         /// when this direct method request is made.
         /// </summary>
+        [JsonIgnore]
+        public TimeSpan ConnectionTimeout { get; set; }
+
+        [JsonProperty("responseTimeoutInSeconds", NullValueHandling = NullValueHandling.Ignore)]
+        internal int? ResponseTimeoutInSeconds => ResponseTimeout <= TimeSpan.Zero
+            ? null
+            : (int)ResponseTimeout.TotalSeconds;
```
Comment (drwill-ms) [inline iothub/service/src/DirectMethod/DirectMethodRequest.cs]: Maybe let them send invalid values. We did before.

Then, at least, they get an error message.

File: iothub/service/src/DirectMethod/DirectMethodRequest.cs [hunk 13]
```diff
@@ -18,58 +18,65 @@ public class DirectMethodRequest
...
+            ? null
+            : (int)ConnectionTimeout.TotalSeconds;
 
         [JsonProperty("payload")]
-        internal JRaw Payload { get; set; }
+        internal JRaw JsonPayload { get; set; }
 
         /// <summary>
-        /// Set the serialized JSON payload.
+        /// Get payload as json
```
Comment (drwill-ms) [inline iothub/service/src/DirectMethod/DirectMethodRequest.cs:61]: Get and set the serialized JSON payload.

File: iothub/service/src/DirectMethod/DirectMethodRequest.cs [hunk 14]
```diff
@@ -18,58 +18,65 @@ public class DirectMethodRequest
...
         /// </summary>
-        public DirectMethodRequest SetPayloadJson(string json)
+        [JsonIgnore]
+        public string Payload
         {
-            if (json == null)
+            get
```
Comment (drwill-ms) [inline iothub/service/src/DirectMethod/DirectMethodRequest.cs:66]: get => (string)JsonPayload.Value;

**Review by drwill-ms — APPROVED** [review]

File: iothub/service/src/DirectMethod/DirectMethodRequest.cs [hunk 15]
```diff
@@ -41,34 +41,27 @@ public class DirectMethodRequest
...
         /// </remarks>
         [JsonIgnore]
-        public TimeSpan ConnectionTimeout { get; set; }
+        public TimeSpan? ConnectionTimeout { get; set; }
 
         [JsonProperty("responseTimeoutInSeconds", NullValueHandling = NullValueHandling.Ignore)]
-        internal int? ResponseTimeoutInSeconds => ResponseTimeout <= TimeSpan.Zero
-            ? null
-            : ResponseTimeout.Seconds;
+        internal int? ResponseTimeoutInSeconds => ResponseTimeout?.Seconds;
```
Comment (drwill-ms) [inline iothub/service/src/DirectMethod/DirectMethodRequest.cs:47]: I think you want TotalSeconds

Comment (drwill-ms) [reply iothub/service/src/DirectMethod/DirectMethodRequest.cs:47]: <https://www.tutorialspoint.com/difference-between-timespan-seconds-and-totalseconds>

Comment (timtay-microsoft) [reply iothub/service/src/DirectMethod/DirectMethodRequest.cs:47]: Nice catch. That's not very intuitive design by Timespan :/

**Pull request merged**